Improvement: Rendering snippets using vs code's virtual document prov…

taksuparth opened the pull request:
**About Pull Request**

> **What I Have done?**

As requested by vscode-snippet project owner ( @mre ), this commit aims to convert manual rendering of snippets to rendering of snippets using vs code's virtual document provider.

Related: #25

> **What We have to test?**

I need some clarification to test locally (original project Doesn't have an explanation on how We could up the project and test)

taksuparth commented:
Hey @mre I have tested it locally and it was generating new pages, I checked for set, for, map, hello world in different languages and they were returning correct snippets. But, I don't clearly get the next and previous functionality. They were also generating new snippets of that type but different implementation.

mre commented:
I tested your code and it works perfectly for me. Really great effort! 🎉 

> They were also generating new snippets of that type but different implementation.

Sometimes an answer does not 100% match the user's expectations and that's when an alternative answer comes in handy.
The "next" and "previous" functionality allows users to paginate through these alternative answers. As you said, the snippets answer the same question, but the code is different. Hope that makes it a little clearer. But as I said, your code works great. We don't have to change anything there.

One minor thing we could do is add the snippet number if it isn't the first answer (so when we called "next answer"). For example, if we show the second snippet, the title could be `[language] foo bar (2)` and the third snippet would be`[language] foo bar (3)`. This way, users would be able to find their snippets more easily. In the long run, we could also make the snippet titles configurable.

Another thing is that "toggle comments" doesn't always work for me. When I have both the commented and the non-commented version open, it doesn't switch between the two when I call `Snippet: Toggle Comments`. Maybe that's a limitation of VSCode because both files have the same name, but I'm not sure. It's just a minor thing, though and shouldn't be a blocker for merging.

Apart from that, great work. Thanks.

taksuparth commented:
Oh cool! I guess I can take that up too. Should I make a different MR for that or do you want me to add those changes in this MR only?

mre commented:
Maybe adding the answer number would be great. The rest could be done in a separate MR.

taksuparth commented:
Hey! @mre I was checking my code after rebasing with master and I am constantly getting this error 
```
Error: Cannot find module 'vscode'
    at Function.Module._resolveFilename (module.js:548:15)
    at Function.Module._load (module.js:475:25)
    at Module.require (module.js:597:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/parthshah/projects/vscode-snippet/out/extension.js:3:16)
    at Module._compile (module.js:650:14)
    at Object.Module._extensions..js (module.js:664:10)
    at Module.load (module.js:566:32)
    at tryModuleLoad (module.js:506:12)
    at Function.Module._load (module.js:498:3)
```
tried searching it but couldn’t find any working solution. I tried to run vscode-smaple-extension and they were running fine. Any idea as to why this must be happening ?

mre commented:
Could it be because dependabot updated some dependencies lately? Can you try another `npm install`?

mre commented:
Reference: https://github.com/mre/vscode-snippet/commit/2609fb62d3fca624b7767a1ac1f2aa9334c753be

taksuparth commented:
Still the same error

mre commented on package.json:
```diff
@@ -1,159 +1,159 @@
+  "version": "0.7.0",
+  "publisher": "vscode-snippet",
+  "engines": {
+    "vscode": "^1.39.0"
```
That's the only line I changed in this file. Please ignore the rest as it's just formatting.

mre commented:
Tried to reproduce it locally. For me it complained about the version mismatch between the `vscode` engine and `@types/vscode`.
So I synced the version in your branch and then it started working again for me.

Can you try to delete the `node_modules` folder and then run this command?

```
make update-vsce install package
```

After that, please try again. 🤞

mre commented:
Hey @taksuparth,
did you get a chance to try it? Otherwise the PR looks good from my side and we can merge now.
If I can help, just let me know. 😊

taksuparth commented:
Hey @mre I didn't get time to try it. Checking now.

taksuparth commented:
Now, working for me. It seems the issue was due to improper launch config inside `.vscode -> launch.json`. Need to update it.

taksuparth commented:
Added answer number in the title bar of each snippet.
<img width="1180" alt="Screenshot 2019-10-31 at 1 05 34 AM" src="https://user-images.githubusercontent.com/13138123/67892421-b1dad000-fb7a-11e9-9089-0fd193f0761b.png">
Fixed a bug - `Show next answer` was always showing the next answer in the plain text.
Language identifiers to be used: https://code.visualstudio.com/docs/languages/identifiers

mre commented on src/query.ts:
```diff
@@ -5,8 +5,7 @@ import { cache } from './cache'
     return new Promise((resolve, _reject) => {
-        let window = vscode.window
-        const quickPick = (<any>window).createQuickPick()
+        const quickPick = vscode.window.createQuickPick()
```
Thanks for cleaning that up.

mre commented on src/snippet.ts:
```diff
@@ -47,27 +47,31 @@ export class Snippet {
     }
 
-    async loadNext(): Promise<Response> {
+    getNextAnswerNumber(): number {
```
That's a better abstraction than what we had before because it adheres to the single-responsibility principle. Good work.

mre commented:
Great work.
I tested the new extension thoroughly. Here are some regressions:
* [ ] `Snippet: Show previous answer` will happily show negative answer numbers (-1, -2, and so on).
* [ ]  `Snippet: Find inplace` opens a separate editor, but it should paste the answer at the cursor of the current editor.
* [ ] `Snippet: Toggle comments` only works once: After switching off comments, reactivating them does not work; i.e. it will get stuck on the editor without comments.
* [ ] The setting `Snippet: Verbose` is ignored. As a consequence, answers will always start to be shown with comments even if `Snippet: Verbose` is set to `false`.
* [ ] The setting `Snippet: Open In New Editor` is ignored. As a consequence, answers will always be shown in a new window, even if `Snippet: Open In New Editor` is set to `false`.

All these things are not super critical, but we should fix them before the next release.
@taksuparth do you want to fix those issues before the merge or shall we merge this already and create a separate issue for the regressions?

taksuparth commented:
I will fix these regression issues tonight. Thank you so much @mre for finding these beforehand.

taksuparth commented:
Hey @mre! I am so sorry, I just forgot about this one. I will get this done by Wednesday.

mre commented:
No worries. 😉 I just had a bit of time on my hides and wanted to hack on this.

mre commented:
Hey @taksuparth,
I guess I was able to resolve the final issues before we can merge this.
Can you look over the changes to see if you find anything spurious?
If you find the time, maybe you could also run it and see if everything works for you.

taksuparth commented:
Hey @mre, 
changes looks good! Tried on local machine, and it works flawlessly 👍

mre commented:
This has finally landed. Thanks for spearheading this @taksuparth. 🎉 
Will cut a new release now.
